compiler/src/dotty/tools/dotc/parsing/Parsers.scala

```diff
@@ -1680,21 +1680,16 @@ object Parsers {
     )
 
     def addFlag(mods: Modifiers, flag: FlagSet): Modifiers = {
-      def incompatible(kind: String) = {
-        syntaxError(ModifiersNotAllowed(mods.flags, kind))
-        Modifiers(flag)
-      }
       if (compatible(mods.flags, flag)) mods | flag
-      else flag match {
-        case Trait => incompatible("trait")
-        case Method => incompatible("method")
-        case Mutable => incompatible("variable")
-        case _ =>
-          syntaxError(s"illegal modifier combination: ${mods.flags} and $flag")
-          mods
+      else {
+        syntaxError(ModifiersNotAllowed(mods.flags, getPrintableTypeFromFlagSet(flag)))
+        Modifiers(flag)
       }
     }
 
+    private def getPrintableTypeFromFlagSet(flag: FlagSet) =
+      Map(Trait -> "trait", Method -> "method", Mutable -> "variable").get(flag)
+
```

Comment:
Since this method is only used by `addFlag`, I would nest it in the method like:

```scala
def addFlag(mods: Modifiers, flag: FlagSet): Modifiers = {
  def getPrintableTypeFromFlagSet(flag: FlagSet) =
    Map(Trait -> "trait", Method -> "method", Mutable -> "variable").get(flag)

  if (compatible(mods.flags, flag)) mods | flag
  else {
    syntaxError(ModifiersNotAllowed(mods.flags, getPrintableTypeFromFlagSet(flag)))
    Modifiers(flag)
  }
}
```